File: 1713-dot-product-of-two-sparse-vectors/dot-product-of-two-sparse-vectors.java
class SparseVector {
    int[] nums;
    int length ;
    SparseVector(int[] nums) {
        this.nums = nums;
        this.length = nums.length;
    }
    
	// Return the dotProduct of two sparse vectors
    public int dotProduct(SparseVector vec) {
        int ans = 0;
        for (int i = 0; i < vec.length; i++) {
            if (nums[i] == 0 || vec.nums[i] == 0) {
                continue;
            }
            else {
                ans += nums[i] * vec.nums[i];
            }
        }
        return ans;
    }
}

// Your SparseVector object will be instantiated and called as such:
// SparseVector v1 = new SparseVector(nums1);
// SparseVector v2 = new SparseVector(nums2);
// int ans = v1.dotProduct(v2);

File: 1713-dot-product-of-two-sparse-vectors/dot-product-of-two-sparse-vectors.py
class SparseVector:
    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.non_zeros = []
        for idx, n in enumerate(nums):
            if n != 0:
                self.non_zeros.append([idx, n])

    def dotProduct(self, vec):
        """
        :type vec: 'SparseVector'
        :rtype: int
        """
        res = 0
        p = 0
        q = 0
        while p < len(self.non_zeros) and q < len(vec.non_zeros):
            if self.non_zeros[p][0] == vec.non_zeros[q][0]:
                res += self.non_zeros[p][1] * vec.non_zeros[q][1]
                p += 1
                q += 1
            elif self.non_zeros[p][0] < vec.non_zeros[q][0]:
                p += 1
            else:
                q += 1
        return res